// @iduty/app/dist/src/Cache/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.$clearStorage = exports.$removeStorage = exports.$getStorage = exports.$setStorage = exports.$clearSession = exports.$removeSession = exports.$getSession = exports.$setSession = void 0;
const Toast_1 = __importDefault(require("../Toast"));
const $setSession = (key, value) => {
    if (!key) {
        (0, Toast_1.default)('请检查输入的参数key');
        return;
    }
    if (value === void (0)) {
        (0, Toast_1.default)('请检查输入的参数value');
        return;
    }
    window.sessionStorage.setItem(key, JSON.stringify(value));
};
exports.$setSession = $setSession;
const $getSession = (key) => {
    if (!key) {
        (0, Toast_1.default)('请检查输入的参数key');
        return;
    }
    const res = window.sessionStorage.getItem(key);
    if (res === null || res === void (0)) {
        return null;
    }
    else {
        return JSON.parse(res);
    }
};
exports.$getSession = $getSession;
const $removeSession = (key) => {
    if (!key) {
        (0, Toast_1.default)('请检查输入的参数key');
        return;
    }
    window.sessionStorage.removeItem(key);
};
exports.$removeSession = $removeSession;
const $clearSession = () => {
    window.sessionStorage.clear();
};
exports.$clearSession = $clearSession;
const $setStorage = (key, value, expires) => {
    if (!key) {
        (0, Toast_1.default)('请检查输入的参数key');
        return;
    }
    if (value === void (0)) {
        (0, Toast_1.default)('请检查输入的参数value');
        return;
    }
    let item = { value };
    if (expires) {
        const exp = parseInt(expires, 10);
        item['expires'] = Date.now() + exp * 1000 * 60;
    }
    window.localStorage.setItem(key, JSON.stringify(item));
};
exports.$setStorage = $setStorage;
const $getStorage = (key) => {
    if (!key) {
        (0, Toast_1.default)('请检查输入的参数key');
        return;
    }
    let res = window.localStorage.getItem(key);
    if (res === null || res === void (0)) {
        res = null;
    }
    else {
        res = JSON.parse(res);
        if (res.hasOwnProperty('expires')) {
            if (res['expires'] > Date.now()) {
                res = res['value'];
            }
            else {
                (0, exports.$removeStorage)(key);
                res = null;
            }
        }
        else {
            res = res['value'];
        }
    }
    return res;
};
exports.$getStorage = $getStorage;
const $removeStorage = (key) => {
    if (!key) {
        (0, Toast_1.default)('请检查输入的参数key');
        return;
    }
    window.localStorage.removeItem(key);
};
exports.$removeStorage = $removeStorage;
const $clearStorage = () => {
    window.localStorage.clear();
};
exports.$clearStorage = $clearStorage;
