// @iduty/app/dist/src/Navigate/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.$toLogin = exports.$toBack = exports.$toReplace = exports.$toJump = void 0;
const qs_1 = __importDefault(require("qs"));
const Env_1 = __importDefault(require("../Env"));
const Config_1 = __importDefault(require("../Config"));
const Cache_1 = require("../Cache");
const getPath = (url, params = {}) => {
    let __path = '';
    let __keys = '';
    if (params && Object.keys(params).length) {
        __keys = `?${qs_1.default.stringify(Object.assign({}, params), { encode: true })}`;
    }
    if (/^\/pages\/other/.test(url) || /^pages\/other/.test(url)) {
        __path = url + __keys;
    }
    else {
        if (!/^http/.test(url)) {
            __path = Config_1.default.domain + (!/^\//.test(url) ? '/' + url : url) + __keys;
        }
        if (Env_1.default.isJban)
            __path = '/pages/other/webview/index?url=' + encodeURIComponent(__path);
    }
    return __path;
};
const $toJump = (url, params = {}) => {
    const __path = getPath(url, params);
    if (Env_1.default.isJban) {
        uni.navigateTo({
            url: __path
        });
    }
    else {
        window.location.href = __path;
    }
};
exports.$toJump = $toJump;
const $toReplace = (url, params = {}) => {
    const __path = getPath(url, params);
    if (Env_1.default.isJban) {
        uni.redirectTo({
            url: __path
        });
    }
    else {
        if (Env_1.default.isAndroid) {
            window.history.replaceState({}, '', __path);
            window.location.reload();
        }
        else {
            window.location.replace(__path);
        }
    }
};
exports.$toReplace = $toReplace;
const $toBack = (params) => {
    const opts = {
        step: 1
    };
    if (params && params.step)
        opts.step = params.step;
    if (Env_1.default.isJban) {
        uni.navigateBack({
            delta: opts.step,
        });
    }
    else {
        window.history.go(-opts.step);
    }
};
exports.$toBack = $toBack;
const $toLogin = () => {
    (0, Cache_1.$removeStorage)('$$token');
    if (Env_1.default.isJban) {
        (0, exports.$toJump)('/pages/other/login1/login2');
    }
    else {
        (0, exports.$toJump)('/business/#/user/login');
    }
};
exports.$toLogin = $toLogin;
