// @iduty/ui/dist/Confirm/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const app_1 = require("@iduty/app");
const Box_1 = __importDefault(require("../Box"));
;
exports.default = (params) => {
    let opts = {
        title: '',
        message: '',
    };
    if (typeof params !== 'object') {
        opts.message = params;
    }
    else {
        opts = Object.assign({}, params);
    }
    const config = {
        type: 'confirm',
        children: opts.message,
        onSubmit: () => {
            var _a;
            app_1.$iduty.destroy('iDutyConfirm');
            (_a = opts === null || opts === void 0 ? void 0 : opts.onSubmit) === null || _a === void 0 ? void 0 : _a.call(opts);
        },
        onCancel: () => {
            var _a;
            app_1.$iduty.destroy('iDutyConfirm');
            (_a = opts === null || opts === void 0 ? void 0 : opts.onCancel) === null || _a === void 0 ? void 0 : _a.call(opts);
        }
    };
    if (opts.title)
        config.title = opts.title;
    if (opts.submitBtn)
        config.submitBtn = opts.submitBtn;
    if (opts.cancelBtn)
        config.cancelBtn = opts.cancelBtn;
    if (opts.className)
        config.className = opts.className;
    app_1.$iduty.set('iDutyConfirm', react_1.default.createElement(Box_1.default, Object.assign({}, config)));
};
